README.md
WP Subtitle
===========

Add subtitles (subheadings) to your pages, posts or custom post types.

The WP Subtitle plugin allows your pages and posts to contain a subtitle.  Also called a sub-heading, this this short line of text is meant to appear beneath a post's (or page's) title, but can be inserted in your template wherever you choose.

The subtitle can be inserted into your theme template files (or plugin) using the following API:

### Display The Subtitle

All parameters are optional. If 'post_id' is omitted then the current post ID in the loop is used.

PHP Code:

```
do_action( 'plugins/wp_subtitle/the_subtitle', array(
    'before'        => '<p class="subtitle">',
    'after'         => '</p>',
    'post_id'       => get_the_ID(),
    'default_value' => ''
) );
```

Output:

`<p class="subtitle">My Post Subtitle</p>`

### Get The Subtitle

All parameters are optional. If 'post_id' is omitted then the current post ID in the loop is used.

A default value can be supplied as the second parameter for `apply_filters`. This will be used if the post does not have a subtitle. Leave as an empty string to return an empty string if the post does not have a subtitle.

PHP Code:

```
$subtitle = apply_filters( 'plugins/wp_subtitle/get_subtitle', '', array(
    'before'  => '<p class="subtitle">',
    'after'   => '</p>',
    'post_id' => get_the_ID()
) );
```

Result:

`$subtitle = '<p class="subtitle">My Post Subtitle</p>'`

### Parameters

The array of arguments accepted for the `plugins/wp_subtitle/the_subtitle` action and `plugins/wp_subtitle/get_subtitle` filter are:

**before**  
*(string)* Text to place before the subtitle if one exists. Defaults to an empty string.

**after**  
*(string)* Text to place after the subtitle if one exists. Defaults to to an empty string.

**post_id**  
*(integer)* Post, page or custom post type ID.

**default_value**  
*(string)* Only used by the `plugins/wp_subtitle/the_subtitle` action, allows you to specify a default subtitle to display if the post does not have one. For the `plugins/wp_subtitle/get_subtitle` filter the second parameter of `apply_filters` should be used instead. Defaults to to an empty string.

### Post Type Support

By default, subtitle are supported by both posts and pages. To add support for custom post types add teh following to your theme functions file or plugin:

`add_post_type_support( 'my_post_type', 'wps_subtitle' )`

### WooCommerce Plugin Support

Subtitles can automatically be added to your WooCommerce products without needing to make template changes. In the admin go to WooCommerce > Settings > Products where you can choose to:

 - Enable Product Subtitles
 - Display the subtitle on single product pages
 - Display the subtitle on product archives (category pages)

### Yoast SEO and SEOPress Plugin Support

The plugin allows you to include the subtitle in your meta titles and descriptions via the [Yoast SEO](https://wordpress.org/plugins/wordpress-seo/) and [SEOPress](https://wordpress.org/plugins/wp-seopress/) plugins.

Similar to the Yoast `%%title%%` placeholder which inserts the post title, you can use `%%wps_subtitle%%`.

There are also addition placeholders and filters to allow to to customize seperators for the subtitle.

For more information, [view the SEO support documentation here](https://github.com/benhuson/wp-subtitle/wiki/Yoast-SEO-Plugin-Support).

Installation
------------

1. Upload the WP Subtitle plugin to your WordPress site in the `/wp-content/plugins` folder or install via the WordPress admin.
1. Activate it from the Wordpress plugin admin screen.
1. Use the API to display the subtitle in your theme.

For full details on the API and how to display the subtitle, [view the documentation here](https://github.com/benhuson/wp-subtitle/wiki).

Frequently Asked Questions
--------------------------

__What does WP Subtitle do?__  

The plugin adds a Subtitle field when editing posts, pages or custom post types. The subtitle is stored as a custom field (post meta data) and can be output using API actions and filters.

__Where does WP Subtitle store the subtitles?__  

All subtitles are stored as post meta data. Deactivating this plugin will not remove those fields.

__Compatibility with WordPress 5.0+__  

In the new editor in WordPress 5.0 the subtitle is editable via a panel in the sidebar (like excerpts).

__How do I add the subtitle to my pages?__  

Refer to [the documentation](https://github.com/benhuson/wp-subtitle/wiki).

__How do I add support for custom post types?__  

To add support for custom post types add the following to your theme functions file or plugin:

```
function my_wp_subtitle_page_part_support() {
	add_post_type_support( 'my_post_type', 'wps_subtitle' );
}
add_action( 'init', 'my_wp_subtitle_page_part_support' );
```

__Where can I get help?__  

Please post support requests and questions in the [WordPress.org Support](http://wordpress.org/support/plugin/wp-subtitle) forum.

__How should I report a bug?__  

Please submit bugs/errors directly to the [GitHub Issues](https://github.com/benhuson/wp-subtitle/issues) list.

__How can I contribute code?__  

The plugin is [hosted on GitHub](https://github.com/benhuson/wp-subtitle) and pull requests are welcome.

Upgrade Notice
--------------

### 3.4.1
Resolve XSS issue by improving sanitization when saving subtitle custom field.

### 3.4
Added support for the SEOPress plugin and updating the subtitle via the REST API.

### 3.3.1
Fixed broken closing H2 tag for WooCommerce subtitle.

### 3.3
New API for getting and displaying the subtitle using `do_action( 'plugins/wp_subtitle/the_subtitle' )` and `apply_filters( 'plugins/wp_subtitle/get_subtitle', '' )`. Please see the documentation.

### 3.2
Fix WordPress 5.0 compatibility.

### 3.1
WooCommerce compatibility: Go to `WooCommerce > Settings > Products > Display` for settings. Yoast SEO compatibility: Added `%%wps_subtitle%%` placeholders. Gutenberg compatibility: Use metabox UI.

### 3.0
Make `wps_subtitle` available via WordPress REST API.

### 2.9.1
Fix preview not rendering correct template and other post meta.

### 2.9
Add support for revisions and fix special character encoding.

### 2.8.1
Fix PHP warning - `get_admin_subtitle_value()` should be declared static.

### 2.8
Allow subtitle to contain HTML (same as main post title ) and add `wps_default_subtitle` filter.

### 2.7.1
Fix incorrect post ID reference preventing subtitle from saving.

### 2.7
Trim subtitle and wptexturize() by default.

### 2.6
Added quick edit support for subtitle. Security Update: Sanitize `$_REQUEST` and `$_GET` when establishing post type in the admin.

### 2.5
Add [wp_subtitle] shortcode. Do not use variable for textdomain - causes issues for parsers.

### 2.4.1
Fix PHP notice warning on 404 error page.

### 2.4
Add subtitle admin column.

### 2.3.1
Security Update: Ensure subtitles are sanitized when saving.

### 2.3
Prevent subtitle fields from displaying on unsupported post types and fix issue with quotes in subtitles.

### 2.2
Subtitle field moved to below title field (only in WordPress 3.5+)

### 2.1
Fixed static method warnings and only load admin functionality when needed.

### 2.0
Added custom post type support and support for more recent versions of WordPress.

### 1.0
Initial release.

Changelog
---------

View a list of all plugin changes in [CHANGELOG.md](https://github.com/benhuson/wp-subtitle/blob/master/CHANGELOG.md).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [3.4.1] - 2022-04-21

### Security
- Resolve XSS issue by improving sanitization when saving subtitle custom field.

## [3.4] - 2020-01-31

### Changed
- Pass the current post object to the `wps_subtitle_field_placeholder` filter. Props [Dominik Schilling](https://github.com/ocean90).

### Added
- Added support for the SEOPress plugin. Props @chriselkins.
- You can now update the subtitle via the REST API. Props @chriselkins.

## [3.3.1] - 2020-01-29

### Fixed
- Fixed broken closing H2 tag for WooCommerce subtitle. Props @faktorvier.

## [3.3] - 2019-09-04

### Changed
- New API for displaying the subtitle using `do_action( 'plugins/wp_subtitle/the_subtitle' )`.
- New API for getting the subtitle using `apply_filters( 'plugins/wp_subtitle/get_subtitle', '' )`.

### Fixed
- Admin column title now matches the meta box title if altered using the `wps_meta_box_title` filter.

## [3.2] - 2018-12-10

### Changed
- Fix WordPress 5.0 compatibility - check `use_block_editor_for_post_type`.

## [3.1] - 2018-09-04

### Added
- `%%wps_subtitle%%` placeholders for Yoast SEO compatibility.
- WooCommerce compatibility. Go to `WooCommerce > Settings > Products > Display` for settings.
- `wps_subtitle_field_position` filter to position admin field `after_title`, `before_title` or otherwise in a meta box.
- Use metabox UI if editing in Gutenberg.

## [3.0] - 2017-09-05

### Added
- Make `wps_subtitle` available via WordPress REST API.
- Added `wps_subtitle_field_position` filter to show subtitle admin field `before_title`, `after_title` or in meta box.

## [2.9.1] - 2017-06-01

### Fixed
- Fix preview not rendering correct template and other post meta.

## [2.9] - 2017-05-03

### Added
- Add support for post revisions. Props [Fabian Marz](https://github.com/fabianmarz).

### Fixed
- As of WordPress 4.3 no need to `esc_attr()` AND `htmlentities()` - can mess up special characters.

## [2.8.1] - 2016-09-14

### Fixed
- Fix PHP warning - `get_admin_subtitle_value()` should be declared static.

## [2.8] - 2016-09-07

### Added
- Add `wps_default_subtitle` filter.

### Changed
- Allow subtitle to contain HTML (same as main post title ).
- Use `WP_Subtitle` class to validate saving of subtitle in the admin.

## [2.7.1] - 2016-08-05

### Fixed
- Fix incorrect post ID reference preventing subtitle from saving.

## [2.7] - 2016-08-04

### Changed
- Trim subtitle by default.
- Apply wptexturize() on subtitle.
- Use WP_Subtitle class to manage post subtitle.

## [2.6] - 2015-12-08

### Security
- Sanitize `$_REQUEST` and `$_GET` when establishing post type in the admin.

### Added
- Added quick edit support for subtitle. Props [Fabian Marz](https://github.com/fabianmarz) and [sun](https://github.com/sun).
- Allow subtitle post meta key to be filtered using `wps_subtitle_key`.
- Add German translation. Props [hatsumatsu](https://github.com/hatsumatsu).

## [2.5] - 2015-08-19

### Added
- Add [wp_subtitle] shortcode.

### Fixed
- Do not use variable for textdomain - causes issues for parsers.

### Changed
- Declare methods as public or private.

## [2.4.1] - 2015-06-09

### Fixed
- Fix PHP notice warning on 404 error page. Props Jay Williams.

### Changed
- Add a little space above subtitle field when below title field in admin.

## [2.4] - 2015-04-28

### Added
- Add subtitle admin column.

## [2.3.2] - 2015-02-10

### Fixed
- Show subtitle admin field when adding new post. Props Gabriel Doty.

## [2.3.1] - 2014-10-03

### Security
- Ensure subtitles are sanitized when saving.

## [2.3] - 2014-09-05

### Fixed
- Prevent subtitle fields from displaying on post types for which support has not been added using add_post_type_support(). Previously the fields were displayed but the subtitle would not be saved.
- Escape subtitle admin field value - fixes issues with subtitles with quotes.

## [2.2] - 2014-07-02

### Added
- Added 'wps_subtitle_use_meta_box' filter to allow the edit field to be displayed in a meta box (the old way).

### Changed
- Moved subtitle field from meta box to below title field in WordPress 3.5+ (props Tor Morten)

## [2.1] - 2014-03-12

### Added
- Ready for translation - .pot file added.
- Added deprecated function warnings if WP_DEBUG enabled.

### Fixed
- Fix static method warnings.

### Changed
- Only include admin functionality when needed.

## [2.0.1] - 2013-09-18

### Fixed
- Use `<?php` instead of just `<?`.

### Changed
- Break out some of the code into separate functions.

## [2.0] - 2013-07-29

### Added
- Added custom post type support - use add_post_type_support( '{post_type}', 'wps_subtitle' ).
- Added 'wps_meta_box_title' filter.
- Added 'wps_subtitle' filter.
- Added 'wps_subtitle_field_description' filter.

### Fixed
- Fixed bug in more recent versions of WordPress.

## [1.0] - 2013-07-27

### Added
- First version.

[Unreleased]: https://github.com/benhuson/wp-subtitle/compare/3.4...HEAD
[3.3.1]: https://github.com/benhuson/wp-subtitle/compare/3.3.1...3.4
[3.3.1]: https://github.com/benhuson/wp-subtitle/compare/3.3...3.3.1
[3.3]: https://github.com/benhuson/wp-subtitle/compare/3.2...3.3
[3.2]: https://github.com/benhuson/wp-subtitle/compare/3.1...3.2
[3.1]: https://github.com/benhuson/wp-subtitle/compare/3.0...3.1
[3.0]: https://github.com/benhuson/wp-subtitle/compare/2.9.2...3.0
[2.9.1]: https://github.com/benhuson/wp-subtitle/compare/2.9...2.9.1
[2.9]: https://github.com/benhuson/wp-subtitle/compare/2.8.1...2.9
[2.8.1]: https://github.com/benhuson/wp-subtitle/compare/2.8...2.8.1
[2.8]: https://github.com/benhuson/wp-subtitle/compare/2.7.1...2.8
[2.7.1]: https://github.com/benhuson/wp-subtitle/compare/2.7...2.7.1
[2.7]: https://github.com/benhuson/wp-subtitle/compare/2.6...2.7
[2.6]: https://github.com/benhuson/wp-subtitle/compare/2.5...2.6
[2.5]: https://github.com/benhuson/wp-subtitle/compare/2.4.1...2.5

admin/admin.php
<?php

/**
 * @package     WP Subtitle
 * @subpackage  Admin
 */

add_action( 'plugins_loaded', array( 'WPSubtitle_Admin', '_setup' ) );

class WPSubtitle_Admin {

	/**
	 * Setup
	 *
	 * @since  2.2
	 * @internal
	 */
	public static function _setup() {

		// Language
		load_plugin_textdomain( 'wp-subtitle', false, dirname( WPSUBTITLE_BASENAME ) . '/languages' );

		add_action( 'admin_init', array( 'WPSubtitle_Admin', '_admin_init' ) );
		add_action( 'post_updated', array( 'WPSubtitle_Admin', '_save_post' ), 9 );
		add_action( 'save_post', array( 'WPSubtitle_Admin', '_save_post' ) );
		add_action( 'admin_enqueue_scripts', array( 'WPSubtitle_Admin', '_add_admin_scripts' ) );

		add_filter( '_wp_post_revision_fields', array( 'WPSubtitle_Admin', '_wp_post_revision_fields' ), 9 );
		add_action( 'wp_restore_post_revision', array( 'WPSubtitle_Admin', 'wp_restore_post_revision' ), 10, 2 );

	}

	/**
	 * Admin Init
	 *
	 * @since  2.3
	 * @internal
	 */
	public static function _admin_init() {

		global $pagenow;

		// Get post type
		$post_type = '';

		if ( isset( $_REQUEST['post_type'] ) ) {
			$post_type = sanitize_text_field( $_REQUEST['post_type'] );
		} elseif ( isset( $_GET['post'] ) ) {
			$post_type = get_post_type( absint( $_GET['post'] ) );
		} elseif ( in_array( $pagenow, array( 'post-new.php', 'edit.php' ) ) ) {
			$post_type = 'post';
		}

		// Setup Field / Meta Box
		if ( WPSubtitle::is_supported_post_type( $post_type ) ) {

			$position = self::subtitle_field_position( $post_type );

			if ( 'after_title' == $position ) {
				add_action( 'admin_head', array( 'WPSubtitle_Admin', '_add_admin_styles' ) );
				add_action( 'edit_form_after_title', array( 'WPSubtitle_Admin', '_add_subtitle_field' ) );
			} elseif ( 'before_title' == $position ) {
				add_action( 'admin_head', array( 'WPSubtitle_Admin', '_add_admin_styles' ) );
				add_action( 'edit_form_top', array( 'WPSubtitle_Admin', '_add_subtitle_field' ) );
			} else {
				add_action( 'add_meta_boxes', array( 'WPSubtitle_Admin', '_add_meta_boxes' ) );
			}

			add_filter( 'manage_edit-' . $post_type . '_columns', array( 'WPSubtitle_Admin', 'manage_subtitle_columns' ) );
			add_action( 'manage_' . $post_type . '_posts_custom_column', array( 'WPSubtitle_Admin', 'manage_subtitle_columns_content' ), 10, 2 );
			add_action( 'quick_edit_custom_box', array( 'WPSubtitle_Admin', 'quick_edit_custom_box' ), 10, 2 );

		}

	}

	/**
	 * Add subtitle input to quick edit.
	 *
	 * @since  2.6
	 *
	 * @uses  add_action( 'quick_edit_custom_box' )
	 *
	 * @param  string  $column_name  Column name.
	 * @param  string  $post_type 	 Post type
	 */
	public static function quick_edit_custom_box( $column_name, $post_type ) {

		if ( $column_name !== 'wps_subtitle' ) {
			return;
		}

		wp_nonce_field( 'wp-subtitle', 'wps_noncename' );

		?>
		<fieldset class="inline-edit-col-left inline-edit-col-left-wps-subtitle">
			<div class="inline-edit-col column-<?php echo $column_name; ?>">
				<label>
					<span class="title"><?php esc_html_e( 'Subtitle', 'wp-subtitle' ); ?></span>
					<span class="input-text-wrap"><input type="text" name="wps_subtitle" class="wps_subtitle" value=""></span>
				</label>
			</div>
		</fieldset>
		<?php

	}

	/**
	 * Add subtitle admin column.
	 *
	 * @since  2.4
	 *
	 * @param   array  $columns  A columns
	 * @return  array            Updated columns.
	 */
	public static function manage_subtitle_columns( $columns ) {

		$new_columns = array();

		// Subtitle column after...
		$after_column = '';
		if ( array_key_exists( 'title', $columns ) ) {
			$after_column = 'title';
		} elseif ( array_key_exists( 'name', $columns ) ) {
			$after_column = 'name';
		}

		// Add column
		if ( empty( $after_column ) ) {
			$columns['wps_subtitle'] = __( 'Subtitle', 'wp-subtitle' );
			return $columns;
		}

		// Insert column
		foreach ( $columns as $column => $value ) {
			$new_columns[ $column ] = $value;
			if ( $after_column == $column ) {
				$new_columns['wps_subtitle'] = __( 'Subtitle', 'wp-subtitle' );
			}
		}

		return $new_columns;

	}

	/**
	 * Display subtitle column.
	 *
	 * @since  2.4
	 *
	 * @param  string  $column_name  Column name.
	 * @param  int     $post_id      Post ID
	 */
	public static function manage_subtitle_columns_content( $column_name, $post_id ) {

		if ( $column_name == 'wps_subtitle' ) {

			$subtitle = new WP_Subtitle( $post_id );
			echo '<span data-wps_subtitle="' . esc_attr( $subtitle->get_subtitle() ) . '">' . esc_html( $subtitle->get_subtitle() ) . '</span>';

		}

	}

	/**
	 * Add Admin scripts.
	 *
	 * @since  2.6
	 * @internal
	 */
	public static function _add_admin_scripts( $hook ) {

		if ( 'edit.php' != $hook ) {
			return;
		}

		wp_enqueue_script( 'wps_subtitle', plugins_url( 'js/admin-edit.js', __FILE__ ), false, null, true );

	}

	/**
	 * Add `wps_subtitle` to post revision fields.
	 *
	 * @since     2.9
	 * @internal
	 *
	 * @param  array  $fields  Revision fields.
	 */
	public static function _wp_post_revision_fields( $fields ) {

		$fields['wps_subtitle'] = __( 'Subtitle', 'wp-subtitle' );

		return $fields;

	}

	/**
	 * Restore revisioned `wps_subtitle` value to post.
	 *
	 * @since  2.9
	 *
	 * @param  int  $post_id      Post ID.
	 * @param  int  $revision_id  Revision ID.
	 */
	public static function wp_restore_post_revision( $post_id, $revision_id ) {

		$subtitle = new WP_Subtitle( $post_id );
		$subtitle->restore_post_revision( $revision_id );

	}

	/**
	 * Add Admin Styles
	 *
	 * @since  2.2
	 * @internal
	 */
	public static function _add_admin_styles() {
		?>
		<style>
		#subtitlediv.top {
			margin-top: 5px;
			margin-bottom: 15px;
			position: relative;
		}
		#subtitlediv.top #subtitlewrap {
			border: 0;
			padding: 0;
		}
		#subtitlediv.top #wpsubtitle {
			background-color: #fff;
			font-size: 1.4em;
			line-height: 1em;
			margin: 0;
			outline: 0;
			padding: 3px 8px;
			width: 100%;
			height: 1.7em;
		}
		#subtitlediv.top #wpsubtitle::-webkit-input-placeholder { padding-top: 3px; }
		#subtitlediv.top #wpsubtitle:-moz-placeholder { padding-top: 3px; }
		#subtitlediv.top #wpsubtitle::-moz-placeholder { padding-top: 3px; }
		#subtitlediv.top #wpsubtitle:-ms-input-placeholder { padding-top: 3px; }
		#subtitlediv.top #subtitledescription {
			margin: 5px 10px 0 10px;
		}
		.quick-edit-row-post .inline-edit-col-left-wps-subtitle {
			clear: left;
		}
		</style>
		<?php
	}

	/**
	 * Get Meta Box Title
	 *
	 * @since  2.2
	 *
	 * @uses  apply_filters( 'wps_meta_box_title' )
	 */
	public static function get_meta_box_title( $post_type ) {
		return apply_filters( 'wps_meta_box_title', __( 'Subtitle', 'wp-subtitle' ), $post_type );
	}

	/**
	 * Add Meta Boxes
	 *
	 * @since  2.0
	 * @internal
	 *
	 * @uses  WPSubtitle::get_supported_post_types()
	 * @uses  apply_filters( 'wps_meta_box_title' )
	 * @uses  WPSubtitle_Admin::_add_subtitle_meta_box()
	 */
	public static function _add_meta_boxes() {
		$post_types = WPSubtitle::get_supported_post_types();
		foreach ( $post_types as $post_type ) {

			$positiom = self::gutenberg_supported( $post_type ) ? 'side' : 'normal';

			add_meta_box( 'wps_subtitle_panel',  self::get_meta_box_title( $post_type ), array( 'WPSubtitle_Admin', '_add_subtitle_meta_box' ), $post_type, $positiom, 'high' );

		}
	}

	/**
	 * Add Subtitle Meta Box
	 *
	 * @since  2.0
	 * @internal
	 *
	 * @uses  WPSubtitle::_get_post_meta()
	 * @uses  apply_filters( 'wps_subtitle_field_description' )
	 */
	public static function _add_subtitle_meta_box() {

		global $post;

		$value = self::get_admin_subtitle_value( $post );

		echo '<input type="hidden" name="wps_noncename" id="wps_noncename" value="' . wp_create_nonce( 'wp-subtitle' ) . '" />';

		// As of WordPress 4.3 no need to esc_attr() AND htmlentities().
		// @see  https://core.trac.wordpress.org/changeset/33271
		echo '<input type="text" id="wpsubtitle" name="wps_subtitle" value="' . esc_attr( $value ) . '" autocomplete="off" placeholder="' . esc_attr( apply_filters( 'wps_subtitle_field_placeholder', __( 'Enter subtitle here', 'wp-subtitle' ) ) ) . '" style="width:99%;" />';

		echo apply_filters( 'wps_subtitle_field_description', '', $post );

	}

	/**
	 * Add Subtitle Field
	 *
	 * @since  2.2
	 * @internal
	 *
	 * @uses  WPSubtitle::_get_post_meta()
	 * @uses  apply_filters( 'wps_subtitle_field_description' )
	 */
	public static function _add_subtitle_field() {

		global $post;

		$value = self::get_admin_subtitle_value( $post );

		echo '<input type="hidden" name="wps_noncename" id="wps_noncename" value="' . wp_create_nonce( 'wp-subtitle' ) . '" />';
		echo '<div id="subtitlediv" class="top">';
		echo '<div id="subtitlewrap">';

		// As of WordPress 4.3 no need to esc_attr() AND htmlentities().
		// @see  https://core.trac.wordpress.org/changeset/33271
		echo '<input type="text" id="wpsubtitle" name="wps_subtitle" value="' . esc_attr( $value ) . '" autocomplete="off" placeholder="' . esc_attr( apply_filters( 'wps_subtitle_field_placeholder', __( 'Enter subtitle here', 'wp-subtitle' ) ) ) . '" />';

		echo '</div>';

		// Description
		$description = apply_filters( 'wps_subtitle_field_description', '', $post );
		if ( ! empty( $description ) ) {
			echo '<div id="subtitledescription">' . $description . '</div>';
		}
		echo '</div>';
	}

	/**
	 * Get Admin Subtitle Value
	 *
	 * @since  2.8
	 * @internal
	 *
	 * @param   WP_Post  $post  Post object.
	 * @return  string          Subtitle value.
	 */
	private static function get_admin_subtitle_value( $post ) {

		$subtitle = new WP_Subtitle( $post );

		$value = $subtitle->get_raw_subtitle();

		// Default subtitle if adding new post
		if ( function_exists( 'get_current_screen' ) && empty( $value ) ) {
			$screen = get_current_screen();
			if ( isset( $screen->action ) && 'add' == $screen->action ) {
				$value = $subtitle->get_default_subtitle( $post );
			}
		}

		return $value;

	}

	/**
	 * Save Subtitle
	 *
	 * @since  2.0
	 * @internal
	 *
	 * @uses  WPSubtitle::get_supported_post_types()
	 *
	 * @param  int  $post_id  Post ID or object.
	 */
	public static function _save_post( $post_id ) {

		// Verify if this is an auto save routine. 
		// If it is our form has not been submitted, so we dont want to do anything
		if ( defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE ) {
			return;
		}

		// Verify nonce
		if ( ! self::_verify_posted_nonce( 'wps_noncename', 'wp-subtitle' ) ) {
			return;
		}

		// Check data and save
		if ( isset( $_POST['wps_subtitle'] ) ) {

			$new_value = $_POST['wps_subtitle'];

			$subtitle = new WP_Subtitle( $post_id );

			// Don't save if value not changed
			if ( $subtitle->is_current_subtitle( $new_value ) ) {
				return;
			}

			if ( $subtitle->current_user_can_edit() ) {
				$subtitle->update_subtitle( $new_value );
			}

		}

	}

	/**
	 * Verify Post Edit Capability
	 *
	 * @since        2.0.1
	 * @deprecated   2.7    Use WP_Subtitle->current_user_can_edit() instead.
	 * @internal
	 *
	 * @param   int  $post_id  Post ID.
	 * @return  bool
	 */
	private static function _verify_post_edit_capability( $post_id ) {

		_deprecated_function( '_verify_post_edit_capability()', '2.7', 'WP_Subtitle->current_user_can_edit()' );

		$subtitle = new WP_Subtitle( $post_id );

		return $subtitle->current_user_can_edit();

	}

	/**
	 * Verify Posted Nonce
	 *
	 * @since  2.0.1
	 * @internal
	 *
	 * @param   string  $nonce   Posted nonce name.
	 * @param   string  $action  Nonce action.
	 * @return  bool
	 */
	private static function _verify_posted_nonce( $nonce, $action ) {
		if ( isset( $_POST[ $nonce ] ) && wp_verify_nonce( $_POST[ $nonce ], $action ) ) {
			return true;
		}
		return false;
	}

	/**
	 * edit_form_after_title Supported
	 *
	 * @since  2.2
	 *
	 * @param   string  $post_type  Post type.
	 * @return  bool
	 */
	private static function edit_form_after_title_supported( $post_type = '' ) {
		global $wp_version;

		if ( version_compare( $wp_version, '3.5', '<' ) ) {
			return false;
		}
		return ! apply_filters( 'wps_subtitle_use_meta_box', false, $post_type );
	}

	/**
	 * Gutenberg Supported
	 *
	 * @since  3.1
	 *
	 * @param   string  $post_type  Post type.
	 * @return  bool
	 */
	private static function gutenberg_supported( $post_type = '' ) {

		if ( function_exists( 'use_block_editor_for_post_type' ) && use_block_editor_for_post_type( $post_type ) && ! isset( $_GET['classic-editor'] ) ) {
			return true;
		}

		return false;

	}

	/**
	 * Subtitle Field Position
	 *
	 * @since  3.1
	 *
	 * @param  string  $post_type  Post type.
	 * @param  string              Position.
	 */
	private static function subtitle_field_position( $post_type = '' ) {

		$position = apply_filters( 'wps_subtitle_field_position', 'after_title', $post_type );

		if ( self::gutenberg_supported( $post_type ) ) {
			return '';
		}

		if ( ! self::edit_form_after_title_supported( $post_type ) ) {
			return '';
		}

		return $position;

	}

}
